File: change.go
// Time Complexity : O(nm)
// Space Complexity : O(m)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
initial array of amount +1 size
for each coin overwrite the number of ways for all values till amount with current + ways to make j-coins[i]
only start second loop from coins[i] value as before this all values will remain same
*/

package main

import "fmt"

func change(amount int, coins []int) int {
	if len(coins) == 0 && amount == 0 {
		return 1
	}
	if len(coins) == 0 {
		return 0
	}
	var p = make([]int, amount + 1)
	p[0] = 1

	for i:=0; i< len(coins);i++ {
		for j:=coins[i]; j <= amount; j++  {
			p[j] += p[j-coins[i]]
		}
	}
	return p[amount]
}

func MainChange() {
	fmt.Println(change(5, []int{1,2,5})) //expected 4
}

File: change.js
// Time Complexity : O(S*n) where S is amount and n is coin denominator
// Space Complexity : O(S)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

/**
 * @param {number} amount
 * @param {number[]} coins
 * @return {number}
 */
var change = function(amount, coins) {
    let dp = [new Array(amount+1).fill(0)];
    dp[0][0] = 1;
    for(let i=1; i<coins.length+1; i++) {
        for(let j=1; j<dp[0].length; j++) {
            if(!dp[i]) {
                dp[i] = new Array(dp[0].length);
                dp[i][0] = 1;
            }
            if(coins[i-1]>j) {
                dp[i][j] = dp[i - 1][j];
            } else {
                upIndex = dp[i-1][j];
                prevIndex = dp[i][j-coins[i-1]];
                dp[i][j] = (upIndex+prevIndex);
           }
        }
    }
    let result = dp[dp.length-1][dp[0].length-1];
    return result;
};